### PR TITLE
Tetris

### PR DESCRIPTION
**Pull request by teddykoker**
Also fixed `irrep_compare` which introduced a bug in concatenate.

### CODE REVIEW

**Comment by mitkotak** [comment]
Want me to close out the other PR and make this the README example ?

**Comment by teddykoker** [comment]
I'll merge the other PR later (probably will just move message_example.* -> example.*). Still a lot more to add to this example; so far it only computes messages for first layer. Currently working on the node updates and follow-on layers. I'm adding a `train_tetris.py` (based on [this](https://github.com/e3nn/e3nn-jax/blob/main/examples/tetris_point.py)) script that saves the weights and can be mmap'd into this C version. End result would be matching logits to the jax version for different inputs, invariant under rotation. Once this is working we should have most of the building blocks for other models minus some activation functions and radial basis which should be relatively easy. Might take a few iterations to get the cleanest looking code here though.

Think the tetris code will end up being too large for the readme, so I think it makes sense to just leave a simple example showing off the basic operations.

**Comment by teddykoker** [comment]
Full tetris implementation working now. `./tetris x1 y1 z1 x2 y2 z2 x3 y3 z3 x4 y4 z4` should be matching the output of the JAX implementation in `python run_tetris.py x1 y1 z1 x2 y2 z2 x3 y3 z3 x4 y4 z4`. Slightly dissatisfied with the amount of hardcoding in `model_create()`, however this mostly a limitation of the library currently assuming output irreps of operations are pre-computed.

Nequip should be able to be implemented with only a few more lines of code + implementations for the gate operation and radial basis functions

**Comment by mitkotak** [comment]
Are you open to having an output_irreps generator to take care of the hand-coded-ness in `model_create` ? Can't think of a clean way to do the pre-computation cleanly though with minimal changes

**Comment by teddykoker** [comment]
The solution I was thinking of is having some functions like:

  * `Irreps* irreps_tensor_product(Irreps* i1, Irreps* i2)`: creates irreps of the tensor product between two irreps
  * `Irreps* irreps_concatenate(Irreps* i1, Irreps* i2)`: creates irreps of the concatenation of two irreps
  * `Irreps* irreps_linear(Irreps* input, Irreps* output, bool force_irreps_out)`: creates irreps of the linear operation **removing unmatching irreps if force_irreps_out==false**
  * `int linear_weight_size(Irreps* input, Irreps* output)`: computes the weight matrix size given input and output irreps
 
This way for e.g. you could do something like (in the case of the tetris code):

```c
 
    // manually defined irreps
    Irreps* irreps_hidden = create_irreps("32x0e + 32x0o + 8x1e + 8x1o + 8x2e + 8x2o");
    Irreps* irreps_sh = irreps_create("1x1o + 1x2e + 1x3o");

    model->layers[0].irreps_sh=irreps_sh;

    // manually defined input
    model->layers[0].irreps_sender=irreps_create("1x0e");

    // can compute tensor product irreps 
    model->layers[0].irreps_tp=irreps_tensor_product(model->layers[0].irreps_sh, 
                                                     model->layers[0].irreps_sender);
                                                       
    // can compute output of concat irreps
    model->layers[0].irreps_receiver=irreps_concatenate(model->layers[0].irreps_tp,
                                                        model->layers[0].irreps_sender);

    // model doesn't force irreps out
    model->layers[0].irreps_node=irreps_linear(model->layers[0].irreps_receiver,
                                               irreps_hidden,
                                               false);

    // Can compute linear weight size from irreps too
    model->layers[0].linear_weight_size=linear_weight_size(...)
    model->layers[0].shortcut_weight_size=linear_weight_size(...);

    // this is manually defined anyway
    model->layers[0].denominator=1.5;
```

This could of course then be moved to some function `Layer layer_create(Irreps* irreps_input, Irreps* irreps_hidden, Irreps* irreps_sh)` which could compute all the necessary irreps from there and avoid any code duplication.

Any thoughts on an API like that?

**Comment by mitkotak** [comment]
API sounds good. My concern is how to precompute things since otherwise this might hit the runtime. We can also just benchmark it out.

**Comment by teddykoker** [comment]
The above operations would only need to be run once during `model_create()` so there wouldn't be any overhead for repeated model calls in something like MD. Are you concerned about the startup latency?

**Comment by mitkotak** [comment]
Yup but ehhh we can deal with it when it comes along

**Comment by teddykoker** [comment]
I see. I wouldn't expect these operations to be very costly but I will benchmark them compared to parsing from string when I add them. Were you originally suggesting some sort of code gen?

**Comment by mitkotak** [comment]
Yup but it's fine. Yeah I think it's just me being obsessive.

**Comment by teddykoker** [comment]
Think computing them will actually be faster than parsing the string - see https://github.com/teddykoker/e3nn.c/pull/8#issuecomment-2229871484
